src/Repeater/DefaultRepeaterServer.ts

```diff
@@ -0,0 +1,205 @@
+  }
+
+  public connect(hostname: string) {
+    this._socket = io(this.options.uri, {
+      parser,
+      path: '/api/ws/v1',
+      transports: ['websocket'],
+      timeout: this.options?.connectTimeout,
+      // @ts-expect-error Type is wrong.
+      // Agent is passed directly to "ws" package, which accepts http.Agent
+      agent: this.options.proxyUrl
+        ? new SocksProxyAgent({
+            ...parse(this.options.proxyUrl)
+          })
+        : false,
+      reconnectionAttempts: this.MAX_RECONNECTION_ATTEMPTS,
+      auth: {
+        token: this.options.token,
+        domain: hostname
+      }
+    });
+
+    this.socket.on('connect_error', (error: Error) => {
+      logger.debug(`Unable to connect to the %s host`, this.options.uri, error);
+    });
+
+    this.createPingTimer();
+
+    logger.debug('Event bus connected to %s', this.options.uri);
+  }
+
+  public requestReceived(
+    handler: RepeaterServerEventHandler<
+      RepeaterServerRequestEvent,
+      RepeaterServerRequestResponse
+    >
+  ): void {
+    this.socket.on('request', (payload, callback) => {
+      this.processEventHandler('request', payload, handler, callback);
+    });
+  }
+
+  public reconnectionFailed(
+    handler: RepeaterServerEventHandler<RepeaterServerReconnectionFailedEvent>
+  ): void {
+    this.socket.io.on('reconnect', () => {
+      this.latestReconnectionError = undefined;
+    });
+
+    this.socket.io.on('reconnect_error', (error) => {
+      this.latestReconnectionError = error;
+    });
+
+    this.socket.io.on('reconnect_failed', () => {
+      this.processEventHandler(
+        'reconnection_failed',
+        {
+          error: this.latestReconnectionError
+        },
+        handler
+      );
+    });
+  }
+
+  public errorOccurred(
+    handler: RepeaterServerEventHandler<RepeaterServerErrorEvent, void>
+  ): void {
+    this.socket.on('error', (payload, callback) => {
+      this.processEventHandler('error', payload, handler, callback);
+    });
+  }
+
+  public reconnectionAttempted(
+    handler: RepeaterServerEventHandler<RepeaterServerReconnectionAttemptedEvent>
+  ): void {
+    this.socket.io.on('reconnect_attempt', (attempt) => {
+      this.processEventHandler(
+        'reconnect_attempt',
+        { attempt, maxAttempts: this.MAX_RECONNECTION_ATTEMPTS },
+        handler
+      );
+    });
+  }
+
+  public reconnectionSucceeded(
+    handler: RepeaterServerEventHandler<void, void>
+  ): void {
+    this.socket.io.on('reconnect', () => {
+      this.processEventHandler('reconnect', undefined, handler);
+    });
+  }
+
+  private get socket() {
+    if (!this._socket) {
+      throw new Error(
+        'Please make sure that repeater established a connection with host.'
+      );
+    }
+
+    return this._socket;
+  }
+
+  private processEventHandler<P>(
+    event: string,
+    payload: P,
+    handler: (payload: P) => unknown,
+    callback?: unknown
+  ) {
+    Promise.resolve(handler(payload))
```

Comment:
To enhance code clarity, I suggest to avoid using the Promise API and, instead, use async/await